doc/differentiation/lanczos_smoothing.qbk

```diff
@@ -0,0 +1,161 @@
+[/
+Copyright (c) 2019 Nick Thompson
+Use, modification and distribution are subject to the
+Boost Software License, Version 1.0. (See accompanying file
+LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+]
+
+[section:diff Lanczos Smoothing Derivatives]
+
+[heading Synopsis]
+
+``
+#include <boost/math/differentiation/lanczos_smoothing.hpp>
+
+namespace boost::math::differentiation {
+
+    template <class Real, size_t order=1>
+    class discrete_lanczos_derivative {
+    public:
+        discrete_lanczos_derivative(Real spacing,
+                                    size_t n = 18,
+                                    size_t approximation_order = 3);
+
+        template<class RandomAccessContainer>
+        Real operator()(RandomAccessContainer const & v, size_t i) const;
+
+        template<class RandomAccessContainer>
+        void operator()(RandomAccessContainer const & v, RandomAccessContainer & dvdt) const;
+
+        template<class RandomAccessContainer>
+        RandomAccessContainer operator()(RandomAccessContainer const & v) const;
```

Comment:
Good to see these work on any RandomAccessContainer - can we add a test case for something other than vector<> or else a compile-time concept check using a concept-archetype? I had hoped that Boost.Concepts would have an archetype already written but apparently not :(

BTW the main reason for templating on the container type that I can see, is so that we can apply smoothing to a "view" of a much larger data set.  Idle thought - how much can work with Boost.Range?  Nothing that returns a container obviously, but maybe the rest can?

Reply:
The reason I wanted to template on a `RandomAccessContainer` was only to make sure people didn't have to memcopy out of (say) an Eigen array, which is fairly pedestrian relative to your use. (BTW, I didn't want to put an Eigen dependency in the unit tests, so there is only a check on `std::array` and `boost::numeric::ublas::vector`.)

As for the range access: The filters is pretty long (and gcc generates [bad assembly](https://stackoverflow.com/questions/54384807/slow-vpermpd-instruction-being-generated-why)), so using the indexing syntax `lanczos(v, i)` and looping over it is not that much more expensive than populating the entire vector via `lanczos(v, dvdt)`. That's why I didn't add a `lanczos(v.begin(), v.end(), dvdt.begin())`-style call.

Reply:
Note that if the range-style call is desirable, it's easily added and I'm glad to do it.

Reply:
Actually I wsn't suggesting changing the call: What I meant was we have Boost.Range which implements a subset of a (read-only) containers members and internally is just a wrapper around a pair of iterators  (https://www.boost.org/doc/libs/1_54_0/libs/range/doc/html/range/reference/utilities/iterator_range.html).  I think ranges are promised as a first-class-citizen for C++20 as well.  So I guess what I was wondering was whether:

```
my_smoother(boost::make_iterator_range(x, y), my_container);
```

would work.  It would need a decoupling of the type of the 2 arguments, but a quick scan of your code looked like it might all "just work"?

Mostly thinking out loud yours....

Reply:
Strangely, the iterator ranges have a free function `size()`, but it doesn't work as expected:

```bash
In file included from lanczos_smoothing_test.cpp:16:
../../../boost/math/differentiation/lanczos_smoothing.hpp:500:49: fatal error: call to 'size' is ambiguous
            for(size_t i = m_f.size() - 1; i <= size(v) - m_f.size(); ++i)
                                                ^~~~
lanczos_smoothing_test.cpp:669:5: note: in instantiation of function template specialization 'boost::math::differentiation::discrete_lanczos_derivative<double, 1>::operator()<boost::iterator_range<double *> >' requested here
    lanczos(v1, v2);
    ^
lanczos_smoothing_test.cpp:701:5: note: in instantiation of function template specialization 'test_data_representations<double>' requested here
    test_data_representations<double>();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/iterator:1832:16: note: candidate function [with _Cont = boost::iterator_range<double *>]
constexpr auto size(const _Cont& __c)
               ^
../../../boost/range/size.hpp:55:5: note: candidate function [with SinglePassRange = boost::iterator_range<double *>]
    size(const SinglePassRange& rng)
    ^
1 error generated.
```

If I change all the free size functions to `v.size()`, then it works. I had some reason, which I currently do not recall for using `using std::size; size(v)` rather than `v.size()`.

I'm gonna look up why std::size exists and why I was using it.